storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java

```diff
@@ -0,0 +1,390 @@
+        }
+    }
+
+    @Override
+    public void failRequest(String id) throws AuthorizationException, TException {
+        meterFailRequestCalls.mark();
+        String func = this.idtoFunction.get(id);
+        if (func != null) {
+            Map<String, String> map = new HashMap<>();
+            map.put(DRPCAuthorizerBase.FUNCTION_NAME, func);
+            checkAuthorization(authorizer, map, "failRequest");
+            Semaphore sem = this.idtoSem.get(id);
+            if (sem != null) {
+                this.idtoResult.put(id, new DRPCExecutionException("Request failed"));
+                sem.release();
+            }
+        }
+    }
+
+    protected ConcurrentLinkedQueue<DRPCRequest> acquireQueue(String function) {
+        ConcurrentLinkedQueue<DRPCRequest> reqQueue = requestQueues.get(function);
+        if (reqQueue == null) {
+            reqQueue = new ConcurrentLinkedQueue<DRPCRequest>();
+            requestQueues.put(function, reqQueue);
+        }
+        return reqQueue;
+    }
+
+    private void checkAuthorization(IAuthorizer aclHandler, Map mapping, String operation, ReqContext reqContext) throws AuthorizationException {
+        if (reqContext != null) {
+            ThriftAccessLogger.logAccess(reqContext.requestID(), reqContext.remoteAddress(), reqContext.principal(), operation);
+        }
+        if (aclHandler != null) {
+            if (reqContext == null)
+                reqContext = ReqContext.context();
+            if (!aclHandler.permit(reqContext, operation, mapping)) {
+                Principal principal = reqContext.principal();
+                String user = (principal != null) ? principal.getName() : "unknown";
+                throw new AuthorizationException("DRPC request '" + operation + "' for '" + user + "' user is not authorized");
+            }
+        }
+    }
+
+    private void checkAuthorization(IAuthorizer aclHandler, Map mapping, String operation) throws AuthorizationException {
+        checkAuthorization(aclHandler, mapping, operation, ReqContext.context());
+    }
+
+    // TO be replaced by Common.mkAuthorizationHandler
+    private IAuthorizer mkAuthorizationHandler(String klassname, Map conf) {
+        IAuthorizer authorizer = null;
+        Class aznClass = null;
+        if (StringUtils.isNotBlank(klassname)) {
+            try {
+                aznClass = Class.forName(klassname);
+                authorizer = (IAuthorizer) aznClass.newInstance();
+                if (authorizer != null) {
+                    authorizer.prepare(conf);
+                }
+            } catch (Exception e) {
+                LOG.error("mkAuthorizationHandler failed!", e);
+            }
+        }
+        LOG.debug("authorization class name: {} class: {} handler: {}", klassname, aznClass, authorizer);
+        return authorizer;
+    }
+
+    public Map getConf() {
+        return conf;
+    }
+
+    public static void main(String[] args) throws Exception {
```

Comment:
This main does not work if we are using HTTP.  The servlet is not setup.  We should probably remove this until it actually works.